File: backend-node/src/models/schedule.js
const db = require('../database');
const SettingsController = require('../controllers/settingsController');

class EmployeeSchedule {
  constructor(data) {
    this.id = data.id;
    this.employee_id = data.employee_id;
    this.month = data.month;
    this.year = data.year;
    this.reserved_hours_per_day = data.reserved_hours_per_day || 0;
    this.available_hours_per_month = data.available_hours_per_month || 0;
    this.created_at = data.created_at;
    this.updated_at = data.updated_at;
  }

  static getById(scheduleId) {
    const row = db.fetchOne('SELECT * FROM employee_schedules WHERE id = ?', [scheduleId]);
    return row ? new EmployeeSchedule(row) : null;
  }

  static getEmployeeSchedule(employeeId, month, year) {
    const row = db.fetchOne(
      `SELECT * FROM employee_schedules
       WHERE employee_id = ? AND month = ? AND year = ?`,
      [employeeId, month, year]
    );
    return row ? new EmployeeSchedule(row) : null;
  }

  static getEmployeeYearlySchedule(employeeId, year) {
    return db.fetchAll(`
      SELECT es.*, e.full_name, e.department
      FROM employee_schedules es
      JOIN employees e ON es.employee_id = e.id
      WHERE es.employee_id = ? AND es.year = ?
      ORDER BY es.month
    `, [employeeId, year]);
  }

  static getTeamSchedule(managerId, month, year) {
    return db.fetchAll(`
      SELECT es.*, e.full_name, e.department, e.position, e.id as employee_id
      FROM employee_schedules es
      JOIN employees e ON es.employee_id = e.id
      WHERE e.line_manager_id = ? AND es.month = ? AND es.year = ?
      ORDER BY e.full_name
    `, [managerId, month, year]);
  }

  updateReservedHours(reservedHours) {
    const workHoursPerDay = SettingsController.getWorkHoursPerDay();
    if (reservedHours > workHoursPerDay) {
      throw new Error(`Reserved hours cannot exceed ${workHoursPerDay} hours per day`);
    }

    db.execute(
      `UPDATE employee_schedules
       SET reserved_hours_per_day = ?, updated_at = CURRENT_TIMESTAMP
       WHERE id = ?`,
      [reservedHours, this.id]
    );

    return EmployeeSchedule.getById(this.id);
  }

  getAvailableHours() {
    const settings = SettingsController.getSettings();
    const workHoursPerDay = settings.work_hours_per_day;
    const workDaysPerMonth = settings.work_days_per_month;
    const available = (workHoursPerDay - this.reserved_hours_per_day) * workDaysPerMonth;
    return Math.max(0, available);
  }

  toDict() {
    return {
      id: this.id,
      employee_id: this.employee_id,
      month: this.month,
      year: this.year,
      reserved_hours_per_day: this.reserved_hours_per_day,
      available_hours_per_month: this.getAvailableHours(),
      created_at: this.created_at,
      updated_at: this.updated_at
    };
  }
}

module.exports = EmployeeSchedule;

File: backend/models/schedule.py
from typing import Optional, Dict, Any, List
from datetime import datetime
from database import db
from controllers.settings_controller import SettingsController

class EmployeeSchedule:
    def __init__(self, **kwargs):
        self.id = kwargs.get('id')
        self.employee_id = kwargs.get('employee_id')
        self.month = kwargs.get('month')
        self.year = kwargs.get('year')
        self.reserved_hours_per_day = kwargs.get('reserved_hours_per_day', 0)
        self.available_hours_per_month = kwargs.get('available_hours_per_month', 0)
        self.created_at = kwargs.get('created_at')
        self.updated_at = kwargs.get('updated_at')

    @staticmethod
    def enrich_schedule_dicts(rows):
        """Add dynamically computed available_hours_per_month to raw schedule dicts using live settings."""
        settings = SettingsController.get_settings()
        whpd = settings['work_hours_per_day']
        wdpm = settings['work_days_per_month']
        for row in rows:
            reserved = row.get('reserved_hours_per_day') or 0
            row['available_hours_per_month'] = max(0, (whpd - reserved) * wdpm)
        return rows
    
    @staticmethod
    def get_by_id(schedule_id: int) -> Optional['EmployeeSchedule']:
        query = 'SELECT * FROM employee_schedules WHERE id = ?'
        row = db.fetch_one(query, (schedule_id,))
        return EmployeeSchedule(**row) if row else None
    
    @staticmethod
    def get_employee_schedule(employee_id: int, month: int, year: int) -> Optional['EmployeeSchedule']:
        query = '''
            SELECT * FROM employee_schedules 
            WHERE employee_id = ? AND month = ? AND year = ?
        '''
        row = db.fetch_one(query, (employee_id, month, year))
        return EmployeeSchedule(**row) if row else None
    
    @staticmethod
    def get_employee_yearly_schedule(employee_id: int, year: int) -> List[Dict[str, Any]]:
        query = '''
            SELECT es.*, e.full_name, e.department
            FROM employee_schedules es
            JOIN employees e ON es.employee_id = e.id
            WHERE es.employee_id = ? AND es.year = ?
            ORDER BY es.month
        '''
        rows = db.fetch_all(query, (employee_id, year))
        return EmployeeSchedule.enrich_schedule_dicts(rows)
    
    @staticmethod
    def get_team_schedule(manager_id: int, month: int, year: int) -> List[Dict[str, Any]]:
        query = '''
            SELECT es.*, e.full_name, e.department, e.position, e.employee_id
            FROM employee_schedules es
            JOIN employees e ON es.employee_id = e.id
            WHERE e.line_manager_id = ? AND es.month = ? AND es.year = ?
            ORDER BY e.full_name
        '''
        return db.fetch_all(query, (manager_id, month, year))
    
    def update_reserved_hours(self, reserved_hours: float) -> 'EmployeeSchedule':
        """Update reserved hours and recalculate available hours"""
        work_hours_per_day = SettingsController.get_work_hours_per_day()
        if reserved_hours > work_hours_per_day:
            raise ValueError(f"Reserved hours cannot exceed {work_hours_per_day} hours per day")
        
        query = '''
            UPDATE employee_schedules 
            SET reserved_hours_per_day = ?, updated_at = CURRENT_TIMESTAMP
            WHERE id = ?
        '''
        db.execute(query, (reserved_hours, self.id))
        db.commit()
        
        # Get updated record
        updated = EmployeeSchedule.get_by_id(self.id)
        return updated
    
    def get_available_hours(self) -> float:
        """Calculate available hours for the month"""
        settings = SettingsController.get_settings()
        work_hours_per_day = settings['work_hours_per_day']
        work_days_per_month = settings['work_days_per_month']
        available = (work_hours_per_day - self.reserved_hours_per_day) * work_days_per_month
        return max(0, available)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'id': self.id,
            'employee_id': self.employee_id,
            'month': self.month,
            'year': self.year,
            'reserved_hours_per_day': self.reserved_hours_per_day,
            'available_hours_per_month': self.get_available_hours(),
            'created_at': self.created_at,
            'updated_at': self.updated_at
        }